src/Coverage.jl

```diff
@@ -147,40 +147,43 @@ module Coverage
 
     """
         amend_coverage_from_src!(coverage::Vector{CovCount}, srcname)
+        amend_coverage_from_src!(fc::FileCoverage)
 
-    The code coverage functionality in Julia can miss code lines, which
-    will be incorrectly recorded as `nothing` but should instead be 0
-    This function takes a coverage count vector and a the filename for
-    a Julia code file, and updates the coverage vector in place.
+    The code coverage functionality in Julia only reports lines that have been
+    compiled. Unused functions (or discarded lines) therefore may be incorrectly
+    recorded as `nothing` but should instead be 0.
+    This function takes an existing result and updates the coverage vector
+    in-place to mark source lines that may be inside a function.
     """
-    function amend_coverage_from_src!(coverage::Vector{CovCount}, srcname)
+    amend_coverage_from_src!(coverage::Vector{CovCount}, srcname) = amend_coverage_from_src!(FileCoverage(srcname, read(srcname, String), coverage))
+    function amend_coverage_from_src!(fc::FileCoverage)
         # The code coverage results produced by Julia itself report some
         # lines as "null" (cannot be run), when they could have been run
         # but were never compiled (thus should be 0).
         # We use the Julia parser to augment the coverage results by identifying this code.
         #
         # To make sure things stay in sync, parse the file position
-        # corresonding to each new line
+        # corresponding to each new line
+        content, coverage = fc.source, fc.coverage
         linepos = Int[]
-        open(srcname) do io
+        let io = IOBuffer(content)
             while !eof(io)
                 push!(linepos, position(io))
                 readline(io)
             end
             push!(linepos, position(io))
         end
-        content = read(srcname, String)
         pos = 1
         while pos <= length(content)
-            linestart = minimum(searchsorted(linepos, pos - 1))
+            lineoffset = searchsortedfirst(linepos, pos) - 2
```

Comment:
It's a cleaner way of writing the same thing, and the -2 means we can use `lineoffset` instead of `linestart - 1` below.

Reply:
Sorry, but I don't follow: So of course switching to `searchsortedfirst` makes sense (I was not aware of this function, thanks). But now instead of searching for the "first value in linepos greater than or equal to pos - 1" we search for the "first value in linepos greater than or equal to pos". Moreover, we then subtract 2 from this value; but this is replacing a subtraction of 1 that was used before. So no, I don't think that's the same thing. That said, it is certainly possible that what was there before was wrong (I must admit I forgot the details of my reasoning for writing it), and this is now correct. But I'd like to understand why, or at least have an explicitly statement about this in the commit message / PR, like "also fix incorrect computation of lineoffset", instead of having this change sneaking in alongside a bunch of other unrelated changes...

Reply:
I'd not complain if this was used, which is indeed obviously equivalent to what was there before:
```suggestion
            lineoffset = searchsortedfirst(linepos, pos - 1) - 1
```

Reply:
The former (`minimum(searchsorted())`) returns the one-based count of the occurrences of the pos-1 character in the string (or errors). The change I made returns the two-based line number of the character at `pos` (due to the extra `push!` in the setup loop). The change you suggest is more similar to the original code and just drops the error check.

Reply:
@vtjnash While I don't fully understand your last comment, but I now went in to take a closer look, and now agree that your change fixes a bug, albeit one that does not seem to manifest in practice (at least I couldn't get it to).

In practice, the code change you propose here is equivalent to this (note the change from `first` to `last`):
```suggestion
            lineoffset = searchsortedlast(linepos, pos - 1) - 1
```
So what is going on? the array `linepos` contains for each line the zero-based offset of the first byte of that line. In a file with a single line with code, followed by an empty line, that could be, say, `[ 0, 20 ]`.

We want to convert the one-based byte offset `pos` into a one-based line number; hence we change `pos` to `pos-1` to make it zero-based, matching the array `linepos`. Then we lookup the line: here the current code (as well as the code using `minimum` it replaced) differ from what your new code and also the above equivalent code does: If `pos` refers to the first byte of a line (e.g. `pos=1`), then all return the same. But if `pos` is in the middle of the line (e.g. `pos=5`), then the old code using `minimum` run into an error; and the current code returns line 2; while the new code snippets both produce the correct answer (line 1) in the end.

Personally, I prefer `searchsortedlast(linepos, pos - 1) - 1` because it maps more closely to my mental model: we have the one-based value pos, convert it to zero-based, search the line, which is one-based; and convert it into an offset used to shifting other one-based line numbers, so we must subtract to make it zero-based.

Anyway: In practice, I've never managed to get `Meta.parse` to return a `pos` that did not reference the start of a line, so this has no visible effect anyhow.